src/modules/Game.class.js

```diff
@@ -20,25 +20,233 @@ class Game {
    * If passed, the board will be initialized with the provided
    * initial state.
    */
-  constructor(initialState) {
-    // eslint-disable-next-line no-console
-    console.log(initialState);
+  constructor(
+    initialState = [
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+    ],
+  ) {
+    this.initialState = initialState;
+    this.size = 4;
+    this.score = 0;
+    this.status = 'idle';
+    this.state = initialState.map((row) => row.slice());
   }
 
-  moveLeft() {}
-  moveRight() {}
-  moveUp() {}
-  moveDown() {}
+  /**
+   * Moves tiles in provided direction
+   *
+   * @param {string} callback
+   * Name of corresponding method with an axis
+   * @param {boolean} direction
+   * Reflects the positive or negative direction of the axis
+   */
+
+  addRandomTile() {
+    const emptyCells = [];
+
+    for (let row = 0; row < this.size; row++) {
+      for (let col = 0; col < this.size; col++) {
+        if (this.state[row][col] === 0) {
+          emptyCells.push({ row, col });
+        }
+      }
+    }
+
+    if (emptyCells.length === 0) {
+      return;
+    }
+
+    const { row: randomRow, col: randomCol } =
+      emptyCells[Math.floor(Math.random() * emptyCells.length)];
+
+    this.state[randomRow][randomCol] = Math.random() < 0.9 ? 2 : 4;
+  }
+
+  arraysEqual(a, b) {
+    return a.length === b.length && a.every((val, i) => val === b[i]);
+  }
+
+  moveRow(row, reverse = false) {
+    const workingRow = reverse ? [...row].reverse() : [...row];
+    const withoutZeros = workingRow.filter((n) => n !== 0);
+    const merged = [];
+    let scoreDelta = 0;
+
+    for (let i = 0; i < withoutZeros.length; i++) {
+      if (withoutZeros[i] === withoutZeros[i + 1]) {
+        const mergedValue = withoutZeros[i] * 2;
+
+        merged.push(mergedValue);
+        scoreDelta += mergedValue;
+        i++;
+      } else {
+        merged.push(withoutZeros[i]);
+      }
+    }
+
+    while (merged.length < row.length) {
+      merged.push(0);
+    }
+
+    const resultRow = reverse ? merged.reverse() : merged;
+
+    return { row: resultRow, scoreDelta };
+  }
+
+  transpose(state) {
+    const result = [];
+
+    for (let i = 0; i < state[0].length; i++) {
+      const newRow = [];
+
+      for (let j = 0; j < state.length; j++) {
+        newRow.push(state[j][i]);
+      }
+      result.push(newRow);
+    }
+
+    return result;
+  }
+
+  moveLeft() {
+    if (this.status !== 'playing') {
+      return;
+    }
+
+    let moved = false;
+    let totalScore = 0;
+
+    const newState = this.state.map((row) => {
+      const { row: newRow, scoreDelta } = this.moveRow(row);
+
+      if (!this.arraysEqual(row, newRow)) {
+        moved = true;
+      }
+      totalScore += scoreDelta;
+
+      return newRow;
+    });
+
+    if (moved) {
+      this.state = newState;
+      this.addRandomTile();
+      this.score += totalScore;
+      this.checkWin();
+      this.checkLose();
+    }
+
+    return moved;
```

Comment:
You have similar logic in every method and it is duplicated, try moving it to a helper function and using it everywhere